File: frontend/lib/domain/repositories/job_repository.dart
import '../models/job.dart';
import '../models/job_page.dart';
import '../models/upload_file.dart';

abstract class JobRepository {
  Future<JobPage> listJobs({int page = 1, int limit = 20});

  Future<JobPage> searchJobs({
    required String keyword,
    int page = 1,
    int limit = 20,
  });

  Future<Job> createJob({
    required String title,
    required String description,
    UploadFile? jdFile,
  });

  Future<Job> getJob(String jobId);

  Future<void> deleteJob(String jobId);
}

File: backend/app/repository/job_repository.py
from __future__ import annotations

from typing import Protocol

from app.domain.job import Job
from app.domain.page import Page
from app.database.datasource.job_datasource import JobDatasource

class JobRepository(Protocol):

    async def save(self, job: Job):
        ...

    async def get_job(self, page: int, page_size: int)-> Page[Job]:
        ...

    async def search_jobs(self, keyword: str, page: int, page_size: int) -> Page[Job]:
        ...

    async def get_by_id(self, job_id: str) -> Job | None:
        ...

    async def delete(self, job_id: str):
        ...